tools/c7n_gcp/c7n_gcp/resources/kms.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2019 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from c7n.utils import local_session
+from c7n_gcp.provider import resources
+from c7n_gcp.query import QueryResourceManager, TypeInfo, ChildResourceManager, ChildTypeInfo
+
+
+@resources.register('kms-keyring')
+class KmsKeyRing(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'cloudkms'
+        version = 'v1'
+        component = 'projects.locations.keyRings'
+        enum_spec = ('list', 'keyRings[]', None)
+        scope = None
+        id = 'name'
+
+        @staticmethod
+        def get(client, resource_info):
+            name = 'projects/{}/locations/{}/keyRings/{}' \
+                .format(resource_info['project_id'],
+                        resource_info['location'],
+                        resource_info['key_ring_id'])
+            return client.execute_command('get', {'name': name})
+
+    def get_resource_query(self):
+        if 'query' in self.data:
+            for child in self.data.get('query'):
+                if 'location' in child:
+                    location_query = child['location']
+                    return {'parent': location_query if isinstance(
+                        location_query, list) else [location_query]}
+
+    def _fetch_resources(self, query):
+        super_fetch_resources = QueryResourceManager._fetch_resources
+        session = local_session(self.session_factory)
+        project = session.get_default_project()
+        if query and 'parent' in query:
+            locations = ['projects/{}/locations/{}'.format(project, location)
+                         for location in query['parent']]
+        else:
+            location_resources = session.client(
```

Comment:
rather than inlining the query here, please just use a static data structure that can be referenced across the several resources where this location structure is evident. we really don't want to round trip for this information across many resource types, nor do we want to expose dummy parents resource types just to get it.